src/lib/x509/x509path.cpp

```diff
@@ -154,136 +115,486 @@ check_chain(const std::vector<std::shared_ptr<const X509_Certificate>>& cert_pat
+                   const std::vector<std::set<Certificate_Status_Code>>& ocsp,
+                   bool require_rev_on_end_entity,
+                   bool require_rev_on_intermediates)
+   {
+   for(size_t i = 0; i != results.size() - 1; ++i)
+      {
+      bool had_crl = false, had_ocsp = false;
 
-   if(!cert_path[0]->allowed_usage(usage))
-      res[0].insert(Certificate_Status_Code::INVALID_USAGE);
+      if(i < crl.size() && crl[i].size() > 0)
+         {
+         for(auto&& code : crl[i])
+            {
+            if(code == Certificate_Status_Code::VALID_CRL_CHECKED)
+               {
+               had_crl = true;
+               }
+            results[i].insert(code);
+            }
+         }
+
+      if(i < ocsp.size() && ocsp[i].size() > 0)
+         {
+         for(auto&& code : ocsp[i])
+            {
+            if(code == Certificate_Status_Code::OCSP_RESPONSE_GOOD)
+               {
+               had_ocsp = true;
+               }
 
-   return Path_Validation_Result(res, std::move(cert_path));
+            results[i].insert(code);
+            }
+         }
+
+      if(had_crl == false && had_ocsp == false)
+         {
+         if((require_rev_on_end_entity && i == 0) ||
+            (require_rev_on_intermediates && i > 0))
+            {
+            results[i].insert(Certificate_Status_Code::NO_REVOCATION_DATA);
+            }
+         }
+      }
+   }
+
+}
+
+Path_Validation_Result BOTAN_DLL x509_path_validate(
+   const std::vector<X509_Certificate>& end_certs,
+   const Path_Validation_Restrictions& restrictions,
+   const std::vector<Certificate_Store*>& trusted_roots,
+   const std::string& hostname,
+   Usage_Type usage,
+   std::chrono::system_clock::time_point ref_time,
+   std::chrono::milliseconds ocsp_timeout)
+   {
+   if(end_certs.empty())
+      throw Invalid_Argument("x509_path_validate called with no subjects");
+
+   std::shared_ptr<const X509_Certificate> end_entity(std::make_shared<const X509_Certificate>(end_certs[0]));
+   std::vector<std::shared_ptr<const X509_Certificate>> end_entity_extra;
+   for(size_t i = 1; i < end_certs.size(); ++i)
+      {
+      end_entity_extra.push_back(std::make_shared<const X509_Certificate>(end_certs[i]));
+      }
+
+   std::vector<std::shared_ptr<const X509_Certificate>> cert_path;
+   Certificate_Status_Code path_building_result =
+      PKIX::build_certificate_path(cert_path, trusted_roots, end_entity, end_entity_extra);
+
+   if(path_building_result != Certificate_Status_Code::OK)
+      {
+      return Path_Validation_Result(path_building_result);
+      }
+
+   std::vector<std::set<Certificate_Status_Code>> status =
+      PKIX::check_chain(cert_path, ref_time,
+                        hostname, usage,
+                        restrictions.minimum_key_strength(),
+                        restrictions.trusted_hashes());
+
+   if(path_building_result != Certificate_Status_Code::OK)
+      status[0].insert(path_building_result);
```

Comment:
Thinking about it, returning early makes more sense. That's also what the old implementation does.